### ISSUE TITLE
Compact IRIs and URI schemes (not security)

### ISSUE THREAD

**Issue by azaroth42** [w3c/json-ld-syntax#177]
As an email address is a URI, it would be `@type:@id` in the context.
If I define `mailto` as a prefix, I can manipulate the intended URI in the data. Whether that is for malicious purposes or not. Equally, `content`, `icon`, `data` are all URI schemes.

For example...

```
{
  "@context": {
    "mailto": "http://my.email.stealing.service/thanks/",
    "email": {"@id": "foaf:email_address", "@type": "@id"},
    "Person": "foaf:Person"
  },
  "@id": "http://person.org/",
  "@type": "Person",
  "email": "mailto:person@person.org"
}
```

I know the answer is "don't do that then" ... but ... could we extend the benefits of `@protected` to cover explicitly NOT defining a term:

```
{
 "@context":
  {
    "mailto": {"@id": null, "@protected": true}
  }
}
```

Then contexts could protect the definition of URI schemes they expect to encounter in their data, without wholesale collision prevention of banning *every* scheme.

(From discussion between @azaroth42 and @kasei)

**Comment by gkellogg** [comment]
Perhaps we could define a term with a null `@id` which was protected, which might work. But, to save from an injection attack would require defining all schemes as null.

It might look like this:

```json
{
  "@context": {
    "mailto": {"@id": null, "@protected": true},
    "email": {"@id": "foaf:email_address", "@type": "@id"},
    "Person": "foaf:Person"
  },
  "@id": "http://person.org/",
  "@type": "Person",
  "email": "mailto:person@person.org"
}
```

The IRI Expansion algorithm would probably need a tweak to detect this. Of course, without changing anything, you could get the same effect using the following:

```json
{
  "@context": {
    "mailto": {
      "@id": "http://my.email.stealing.service/thanks/",
      "@protected": true,
      "@prefix": false
    },
    "email": {"@id": "foaf:email_address", "@type": "@id"},
    "Person": "foaf:Person"
  },
  "@id": "http://person.org/",
  "@type": "Person",
  "email": "mailto:person@person.org"
}
```

**Comment by gkellogg** [comment]
Actually, that might not affect IRI expansion, but if we used @rubensworks interpretation, it might, and the expense of backwards compatibility. But, using `@protected` (and `@prefix`) requires `@version: 1.1`, so we should be free to define behavior here.

**Comment by azaroth42** [comment]
`@prefix: false, @protected: true` would mean that the term can *never* be used as a prefix, right? The actual `@id` is irrelevant, as it will never be expanded (unless it's used as a property). That might be sufficient, as the restrictions could be defined in scoped contexts, so as to not pollute the global space for things like `icon` or `content`, but protect them where they are used as a URI scheme.

```
{
  "@context": {
    "email": {
      "@id": "foaf:email_address",
      "@type": "@id",
      "@context": {
        "mailto": {
          "@id": "http://never.used.uri",
          "@prefix": false,
          "@protected": true
        }
     }                                  
  }
}
```

**Comment by gkellogg** [comment]
Currently, it would mean that it can’t be used to create a Compact IRI, but would still be used when expanding. In 1.1 mode, we may want to change that.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-05-17-json-ld#section3-3).

- No actions or resolutions
<details><summary><i class="differentiate">View the transcript</i></summary>
Compact IRIs #177<br/>
Rob Sanderson: link: <a href="https://github.com/w3c/json-ld-syntax/issues/177">https://github.com/w3c/json-ld-syntax/issues/177</a><br/>
Rob Sanderson: if you define a term that’s the same as a URI scheme<br/>
… then you’ll have problems<br/>
… so there are schemes that are very likely to be used in values<br/>
… like mailto: which could be high jacked in this scenario<br/>
… we could fix this by being able to protect URI schemes that are in use<br/>
… by making it possible to set certainly ones to null<br/>
… so, for instance, you could also set them within certain properties–like foaf:mailbox<br/>
… and then protect them to keep upstream ones from overriding mailto:<br/>
… so this means we don’t need to put all possible URI schemes in the spec<br/>
Dave Longley: not enough time to consider it :/<br/>
Rob Sanderson: and context authors can limit the damage to specific spaces where it could be damaging<br/>
… so, something to think about over the week<br/>
</details>

**Comment by dlongley** [comment]
Another option to consider could be a mechanism by which to say "no prefixes"/"no CURIEs" in the instance data (they can only be used in the context). Or an option such that prefixes may only be used when expanding with the vocab flag set to true (wouldn't this cover the above `mailto:` case at least?). Anything with a `:` in it (in the appropriate places) would be interpreted as an absolute IRI instead.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-06-07-json-ld.html#section3-2).

- `RESOLVED:  If <code>@prefix:</code> false works, then we can turn 177 into a best practice. If not, then 177 becomes an issue to improve <code>@prefix</code> to allow it.`
- `ACTION:  test if <code>@prefix:</code> false without <code>@id</code> works as expected (Pierre-Antoine Champin)`
<details><summary><i class="differentiate">View the transcript</i></summary>
3.2. Compact IRIs and URI Schemes<br/>
Ivan Herman: <a href="https://github.com/w3c/json-ld-syntax/issues/177">https://github.com/w3c/json-ld-syntax/issues/177</a><br/>
Rob Sanderson: we discussed this a few weeks ago, but only to point of introducing the issue.<br/>
… hopefully people have thought about it<br/>
… the issue is that IRI schemes that don’t start with // are seen as compact IRIs<br/>
… e.g., icon:… is seen as a compact IRI not a full IRI<br/>
… we could define these as undefined (rather than a prefix)<br/>
… this approach adds a little security without banning all non // schemes<br/>
Ivan Herman: I don’t fully understand - so in your example in the issue, what’s the role of protected<br/>
Rob Sanderson: avoids someone stealing the meaning of prefix (e.g., mailto:)<br/>
Benjamin Young: the goal is to allow continued use of these kinds of IRI prefixes<br/>
Rob Sanderson: having it in the context seems a good solution since it allows you to specify only the scope where you use the scheme rather than everywhere<br/>
Benjamin Young: yes, it is good to scope this<br/>
Pierre-Antoine Champin: it seems to me the protected approach would not be very efficient<br/>
… I’m not sure I fully grasp the use cases, but it seems you still get around protected<br/>
… the only way to really protect against redirection of mailto: is to protect the individual properties (e.g., email)<br/>
Rob Sanderson: so you say that email values not to be expected<br/>
Pierre-Antoine Champin: not sure this requires anything additional since we already have prefix<br/>
… seems a matter of best practice<br/>
Ivan Herman: in the example you have two approaches nul and prefix=false<br/>
Rob Sanderson: probably prefix=false, this works well since we already have prefix<br/>
Ivan Herman: do we?<br/>
Rob Sanderson: issue 76<br/>
Rob Sanderson: ref;: <a href="https://github.com/w3c/json-ld-api/issues/76">https://github.com/w3c/json-ld-api/issues/76</a><br/>
Pierre-Antoine Champin: a more efficient protection would be { "email": { "<code>@id":</code> "<a href="http://example.org/emailAddress",">http://example.org/emailAddress",</a> "<code>@type":</code> "<code>@id",</code> "<code>@protected":</code> true, "<code>@context":</code> { "mailto": { "<code>@prefix":</code> false } } }<br/>
Rob Sanderson: proposal was made to add prefix with 3 allowed values (one of which would be false, as required by issue 177)<br/>
… the issue in doing this was that you need to have an <code>@id,</code> which is why we might still need <code>@id</code> null<br/>
… seems like we shouldn’t need that if prefix: false<br/>
… Pierre-Antoine, does prefix: false already sufficient?<br/>
Pierre-Antoine Champin: not sure<br/>
Ruben Taelman: doesn’t seem to work right now<br/>
Rob Sanderson: but maybe it protects against intentional mischief<br/>
… anyone willing to take action to verify that prefix: false does work or can be made to work?<br/>
Pierre-Antoine Champin: okay, put this action on me<br/>
Action #2: test if <code>@prefix:</code> false without <code>@id</code> works as expected (Pierre-Antoine Champin)<br/>
David I. Lehn: find it a little hard to understand the attack line here<br/>
… if we assume all contexts are a risk, seems a bigger problem<br/>
… protect should be more about extension and override<br/>
Rob Sanderson: we have run into this issue with regard to content being incorrectly expanded<br/>
Benjamin Young: not directly on point, but it feels like a lot of our issues have had to do with managing context intermingling<br/>
… we may need to zoom out a little<br/>
… while there are issues here it seems like its about trust of contexts, ownership of contexts, etc.<br/>
… this fragility that is still there is where we may have issues with security and privacy reviews<br/>
… I sometimes feel like we’re applying pressure to the wrong part of the ecosystem<br/>
… it’s not clear the role and motivation of the actors<br/>
… so maybe we need a broader view<br/>
… would not be an issue if you always trusted the context file you reference<br/>
Rob Sanderson: not certain about that. If you define mailto: you need to undefine it elsewhere in your instance<br/>
Benjamin Young: so this is partly about remixing within instance / context<br/>
Rob Sanderson: I think at least we need to determining how prefix: false works without <code>@id</code> or <code>@id:</code> null<br/>
… need to be able to treat these as resource IRI when you want to<br/>
Ivan Herman: I share bigbluehat’s concerns in a general way<br/>
… for this issue you are asking if prefix: false solves current issue<br/>
… since prefix is a 1.1 key, we need it to work as we want and so if it’s not working that way now, let’s make it work<br/>
Pierre-Antoine Champin: the questions becoming: is it possible (i.e. not over-complicated) to implement <code>@prefix</code> to work like this<br/>
Rob Sanderson: the action should allow us to do this<br/>
Pierre-Antoine Champin: i just realized during discussion we have already introduced redefining a term that looks like an IRI because no useful use case for doing this<br/>
… in a way this desire to limit prefixes that are sometimes used as schemes<br/>
… not clear how critical the use cases are<br/>
… may confuse users to say some IRIs cannot be anything else, but some could be compact IRIs in disguise<br/>
Ivan Herman: does this mean we should have a list of prefixes that are schemes?<br/>
Pierre-Antoine Champin: no, trying hard not to say that<br/>
… current rule is scheme://<br/>
… and this is clear cut, non-ambiguous<br/>
… it might be confusing to do this for non // schemes<br/>
Rob Sanderson: so our approach is to leave this to context authors rather than trying to maintain a universal list<br/>
… this will almost certainly come up in security horizontal review<br/>
… it would be nice to solve or improve it<br/>
… if prefix: false works, we’re done<br/>
Proposed resolution: If <code>@prefix:</code> false works, then we can turn 177 into a best practice. If not, then 177 becomes an issue to improve <code>@prefix</code> to allow it. (Rob Sanderson)<br/>
Rob Sanderson: +1<br/>
Ivan Herman: +1<br/>
Ruben Taelman: +1<br/>
Pierre-Antoine Champin: +1<br/>
Rob Sanderson: otherwise we have to improve prefix: false so it does work<br/>
Tim Cole: +1<br/>
Adam Soroka: +1<br/>
David I. Lehn: +1<br/>
Benjamin Young: +1<br/>
David Newbury: +1<br/>
Resolution #3: If <code>@prefix:</code> false works, then we can turn 177 into a best practice. If not, then 177 becomes an issue to improve <code>@prefix</code> to allow it.<br/>
</details>

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-06-14-json-ld#section3-1).

- No actions or resolutions
<details><summary><i class="differentiate">View the transcript</i></summary>
<b>2019-06-07-action2:</b> test if <code>@prefix:</code> false without <code>@id</code> works as expected (Pierre-Antoine Champin) #87<br/>
Benjamin Young: issue #87 <a href="https://github.com/w3c/json-ld-wg/issues/87">https://github.com/w3c/json-ld-wg/issues/87</a><br/>
Benjamin Young: born from #76 <a href="https://github.com/w3c/json-ld-api/issues/76">https://github.com/w3c/json-ld-api/issues/76</a><br/>
Ivan Herman: original-original: <a href="https://github.com/w3c/json-ld-syntax/issues/177">https://github.com/w3c/json-ld-syntax/issues/177</a><br/>
Benjamin Young: next two issues are related<br/>
Pierre-Antoine Champin: some background: it was possible to define certain URI schemes as prefixes<br/>
… leading to confusing them with the start of CURIs<br/>
… I tried to alter <code>@prefix</code> to avoid this<br/>
… but that turns out not to be possible<br/>
… . so <code>@prefix</code> must be accompanied by <code>@id</code><br/>
… even if you accept ruben’s alteration, it is still not simple<br/>
… <code>@prefix</code> is only used during compaction as a hint to the algo<br/>
… but never it is used in expansion<br/>
… to block interpretation as a CURI<br/>
… so doing this with <code>@prefix</code> would radically change the meaning of <code>@prefix,</code> to add an effect during expansion<br/>
… this thing will come back and bite other people than us<br/>
Adam Soroka: .. users would expect <code>@prefix</code> to work both ways<br/>
Pierre-Antoine Champin: but that may have real effects on implementors<br/>
Gregg Kellogg: yes, <code>@prefixis</code> only meaningful during compaction<br/>
… in 1.0 any prefix could be used<br/>
… so we made some changes<br/>
… including restrictions and a keyword <code>@prefix</code> to overcome them<br/>
… we do run into 1.0 compatibility issues here<br/>
… but this is something we just didn’t complete<br/>
… it’s not a major rewrite<br/>
… now we are forcing the use of expanded term defns for things that used to be simple prefixes<br/>
Dave Longley: +1 to what gregg is saying … it seems to mostly be restricted to very local change in IRI expansion where we need one extra flag check<br/>
Gregg Kellogg: the impact isn’t too bad on the algos<br/>
… we did make changes to limit the use of prefixes during compaction, we should make the same changes for expansion<br/>
Ivan Herman: I thikn this issue arose in a different ticket ^^^^<br/>
… I have said before that we shouldn’t add too many things to the spec<br/>
… do we really have to do anything about #177?<br/>
… does it occur significantly in 1.0 usage?<br/>
… do we need to spend two meetings on it?<br/>
… i don’t think so. my feeling is that we should either close or defer<br/>
… we struggled whether this is really a security issue<br/>
… it can be awkward<br/>
… in several years it did not bite us<br/>
Benjamin Young: we haven’t had <code>@prefix</code> very long either!<br/>
Ivan Herman: but it’s not a matter of just <code>@prefix</code><br/>
Pierre-Antoine Champin: I agree with ivan<br/>
… I’m not convinced that this is a security issue<br/>
… . I’m not convinced that <code>@prefix</code> would be a good way to solve it if it were<br/>
Adam Soroka: .. getting this right with <code>@prefix</code> would be challenging<br/>
Pierre-Antoine Champin: <code>@prefix</code> was introduced for other reasons<br/>
… the way we have it now is counterintuitive<br/>
… we were all talking last week about using it this way without even realizing we couldn’t. that’s not good!<br/>
Gregg Kellogg: when we introduced the restrictions for prefixes, we overlooked expansion<br/>
… we left the job partly done<br/>
Ivan Herman: we are adding features and making the spec more complicated<br/>
Gregg Kellogg: is inconsistency complicated?<br/>
Ivan Herman: we are complicating the spec a few months before we try to get to CR<br/>
Dave Longley: I sympathize with ivan<br/>
… if we consider this a security issue, then <code>@prefix</code> is not how we should address it<br/>
… we could instead introduce a rule that you don’t expand prefixes when <code>@vocab</code> = true<br/>
Gregg Kellogg: I agree with dlongley; that would solve the problem but burn all the fields<br/>
Ivan Herman: I propose that we defer one of the two<br/>
Gregg Kellogg: I don’t think we can roll back prefix<br/>
… it fixes actual errors in 1.0<br/>
Pierre-Antoine Champin: I suggest we make no decision before the next agenda item<br/>
Gregg Kellogg: I can do a speculative PR<br/>
Benjamin Young: let’s go to the next issue, discuss, then come back to this<br/>
</details>

**Comment by gkellogg** [comment]
Is this resolved by w3c/json-ld-api#109 and w3c/json-ld-api#110? Do we need something in the syntax document?

**Comment by pchampin** [comment]
> Is this resolved by w3c/json-ld-api#109 and w3c/json-ld-api#110? 

I think it is.

> Do we need something in the syntax document?

I think it rather belongs to the BP document.
